src/nagini_translation/translators/program.py

```diff
@@ -667,6 +667,269 @@ def create_may_set_predicate(self, ctx: Context) -> 'silver.ast.Predicate':
                                             pos, info)
         return may_set_pred
 
+    def create_adts_domains_and_functions(self, adts: List[PythonClass],
+                           ctx: Context) -> List['silver.ast.domain']:
+        """
+        Translate Algebraic Data Types defined in Python, with classes (sum)
+        and a NamedTuple (product), to a domain in Viper. Further documentation
+        on ADT syntax can be found in ADT.py.
+        """
+
+        def get_cons_args_decl(cons, adt_name, adt_type, pos, info):
+            arguments = []
+            for arg_name, arg_type in cons.fields.items():
+                if arg_type.type.name == adt_name:
+                    argument_type = adt_type
+                else:
+                    argument_type = self.viper.Ref
+                argument = self.viper.LocalVarDecl(arg_name, argument_type, pos, info)
+                arguments.append(argument)
+            return arguments
+
+        def get_cons_args(cons, adt_name, adt_type, pos, info):
+            arguments = []
+            for arg_name, arg_type in cons.fields.items():
+                if arg_type.type.name == adt_name:
+                    argument_type = adt_type
+                else:
+                    argument_type = self.viper.Ref
+                argument = self.viper.LocalVar(arg_name, argument_type, pos, info)
+                arguments.append(argument)
+            return arguments
+
+        def conjoin(eqs, pos, info):
+            if len(eqs) == 1:
+                return eqs[0]
+            else:
+                return self.viper.And(eqs[0], conjoin(eqs[1:], pos, info), pos, info)
+        
+        def disjoin(eqs, pos, info):
+            if len(eqs) == 1:
+                return eqs[0]
+            else:
+                return self.viper.Or(eqs[0], disjoin(eqs[1:], pos, info), pos, info)
+
+        info = self.no_info(ctx)
+        domains = []
+        functions = []
+
+        for adt in adts:
```

Comment:
Whoa, okay, this is a looooong block of code. If at all possible, break this up into smaller functions. Basically, I think you can make a function for each part of the code that has its own comment, and then just call them all from here or something like that.
(I know that my existing code also has some methods that are longer than they should be, btw, I'll try to fix that at some point soon)

Reply:
Done.